internal/validation/gateway/httproute.go

```diff
@@ -0,0 +1,172 @@
+	}
+
+	// validate that no unsupported features are in use
+	if err := validateHTTPRouteFeatures(httproute); err != nil {
+		return false, "httproute spec did not pass validation", err
+	}
+
+	return true, "", nil
+}
+
+// -----------------------------------------------------------------------------
+// Validation - HTTPRoute - Private Functions
+// -----------------------------------------------------------------------------
+
+// validateHTTPRouteListener verifies that a given HTTPRoute is configured properly
+// for a given gateway listener which it is linked to.
+func validateHTTPRouteListener(listener *gatewayv1alpha2.Listener) (bool, string, error) {
+	// verify that the listener supports HTTPRoute objects
+	if listener.AllowedRoutes != nil && // if there are no allowed routes, assume all are allowed
+		len(listener.AllowedRoutes.Kinds) > 0 { // if there are no allowed kinds, assume all are allowed
+		// search each of the allowedRoutes in the listener to verify that HTTPRoute is supported
+		supported := false
+		for _, allowedKind := range listener.AllowedRoutes.Kinds {
+			if allowedKind.Kind == "HTTPRoute" {
+				supported = true
+			}
+		}
+
+		// verify that we found a supported kind
+		if !supported {
+			return false, fmt.Sprintf("HTTPRoute not supported by listener %s", listener.Name), nil
+		}
+	}
+
+	return true, "", nil
+}
+
+// validateHTTPRouteFeatures checks for features that are not supported by this
+// HTTPRoute implementation and validates that the provided object is not using
+// any of those unsupported features.
+func validateHTTPRouteFeatures(httproute *gatewayv1alpha2.HTTPRoute) error {
+	for _, rule := range httproute.Spec.Rules {
+		for _, match := range rule.Matches {
+			// we don't support queryparam matching rules
+			if len(match.QueryParams) != 0 {
+				return fmt.Errorf("queryparam matching is not yet supported for httproute")
+			}
+
+			// we don't support regex path matching rules
+			if match.Path != nil && match.Path.Type != nil && *match.Path.Type == gatewayv1alpha2.PathMatchRegularExpression {
+				return fmt.Errorf("regex path matching is not yet supported for httproute")
+			}
+
+			// we don't support regex header matching rules
+			for _, hdr := range match.Headers {
+				if hdr.Type != nil && *hdr.Type == gatewayv1alpha2.HeaderMatchRegularExpression {
+					return fmt.Errorf("regex header matching is not yet supported for httproute")
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// -----------------------------------------------------------------------------
+// Validation - HTTPRoute - Private Utility Functions
+// -----------------------------------------------------------------------------
+
+// getParentRefForHTTPRouteGateway produces the parentRef that links to a given Gateway
+// object given the HTTPRoute object which links to it.
+func getParentRefForHTTPRouteGateway(httproute *gatewayv1alpha2.HTTPRoute, gateway *gatewayv1alpha2.Gateway) (parentRef *gatewayv1alpha2.ParentRef, err error) {
```

Comment:
this function may be reusable across future route types if we replace `httproute` with `defaultNamespace string`. This will also simplify the parameter list.